File: sorted.java
abstract class Sorted {
    public boolean isSorted(int[] array) {
        int prev = array[0];
    
        for (int i = 0; i < array.length; i++) {
            if (negativeCompare(prev, array[i])) {
                return false;
            }
            prev = array[i];
        }
        return true;
    }
    
    abstract protected boolean negativeCompare(int a, int b);
}

class SortedAscending extends Sorted {
    protected boolean negativeCompare(int a, int b) {
        return a > b;
    }
}

class SortedDescending extends Sorted {
    protected boolean negativeCompare(int a, int b) {
        return a < b;
    }
}

class Solution {
  public static String isSortedAndHow(int[] array) {
      Sorted asc = new SortedAscending();
      Sorted des = new SortedDescending();
  
      if (asc.isSorted(array)) {
          return "yes, ascending";
      }
      
      if (des.isSorted(array)) {
          return "yes, descending";
      }
      
      return "no";
  }
  
}

File: sorted.c
#include <stdbool.h>

bool is_greater(int a, int b)
{
    return a > b;
}

bool is_lower(int a, int b)
{
    return a < b;
}

bool is_sorted_by(int* array, int arrayLength, bool *compare(int, int))
{
    int prev = array[0];
    
    for (int i = 0; i < arrayLength; i++) {
        if (compare(prev, array[i])) {
            printf("soon\n");
            return false;
        }
        prev = array[i];
    }
    
    return true;
}

char* isSortedAndHow(int* array, int arrayLength)
{
    if (is_sorted_by(array, arrayLength, &is_greater)) {
        return "yes, ascending";
    }
    
    if (is_sorted_by(array, arrayLength, &is_lower)) {
        return "yes, descending";
    }
    
    return "no";
}